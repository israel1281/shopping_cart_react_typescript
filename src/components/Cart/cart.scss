.cartContainer {
  background-color: #001f2f;
  position: fixed;
  top: 0;
  width: 550px;
  height: 100%;
  z-index: 99;
  transition: right 0.2s;
  box-sizing: border-box;

  .cartButton {
    padding: 0;
    width: 60px;
    height: 60px;

    border: none;
    font-size: 1.2em;
    color: #040404;
    text-align: center;
    line-height: 50px;
    font-weight: 600;
    top: 0;
    left: 0;
    cursor: pointer;
    position: relative;
    z-index: 2;

    &:focus-visible {
      outline: 3px solid rgb(37, 37, 37);
    }

    &:hover {
      filter: brightness(85%);
    }

    @media only screen and (min-width: 768px) {
      left: -50px;
    }

    .cartIcon {
      width: 45px;
      height: 45px;
      position: relative;
      display: inline-block;
      vertical-align: middle;
      margin-right: 15px;
      color: #050505;
      background-image: url("../../assets/cart-icon.png");
      background-repeat: no-repeat;
      background-size: contain;
      background-position: center;
      background-size: 50%;

      .cartQuantity {
        display: inline-block;
        width: 18px;
        height: 18px;
        color: #001f2f;
        font-weight: bold;
        font-size: 0.6em;
        text-align: center;
        line-height: 18px;
        border-radius: 50%;
        background-color: #eabf00;
        position: absolute;
        bottom: 0;
        right: 5px;
      }
    }
  }

  .cartContent {
    height: 100%;
    overflow-y: scroll;

    @media screen and (min-width: 640px) {
      &::-webkit-scrollbar {
        width: 10px;
        background-color: rgba(0, 0, 0, 0.2);
        padding: 10px;
      }
      &::-webkit-scrollbar-thumb {
        border-radius: 4px;
        background-color: #0c0b10;
      }
    }

    .cartContentHeader {
      color: #ececec;
      box-sizing: border-box;
      text-align: center;
      padding: 45px 0;

      .headerTitle {
        font-weight: bold;
        font-size: 1.2em;
        vertical-align: middle;
      }
    }

    .cartFooter {
      box-sizing: border-box;
      padding: 5%;
      position: absolute;
      bottom: 0;
      width: 100%;
      height: 200px;
      z-index: 2;
      background-color: rgb(252, 252, 252);

      &::before {
        content: "";
        width: 100%;
        height: 20px;
        display: block;
        position: absolute;
        top: -20px;
        left: 0;
        background: linear-gradient(to top, rgba(0, 0, 0, 0.2), transparent);
      }

      .sub {
        width: 20%;
        color: #5b5a5e;
        vertical-align: middle;
        display: inline-block;
      }

      .subPrice {
        width: 80%;
        text-align: right;
        color: #5b5a5e;
        vertical-align: middle;
        display: inline-block;

        .subPriceValue {
          color: black;
          font-size: 22px;
          margin: 0;
        }

        .subPriceInstallment {
          margin: 0;
        }
      }
      .checkoutButton {
        width: 100%;
        border: 0;
        color: #ececec;
        text-transform: uppercase;
        background-color: #f2ab06ef;
        font-weight: 500;
        text-align: center;
        padding: 15px 0;
        margin-top: 40px;
        cursor: pointer;
        outline: none;
        border-radius: 4px;
        transition: background-color 0.2s;

        &:focus-visible {
          outline: 3px solid rgb(54, 54, 54);
        }

        &:hover {
          background-color: #002c41;
        }
      }
    }
  }
}
