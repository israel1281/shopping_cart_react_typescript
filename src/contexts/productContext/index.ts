export { ProductsProvider } from "./ProductContextProvider";
export { default as useProducts } from "./useProducts";
