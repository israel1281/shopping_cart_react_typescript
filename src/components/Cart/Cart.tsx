import formatPrice from "../../utils/formatPrice";
import CartProducts from "./CartProducts";

import { useCart } from "../../contexts/cartContext";

import "./cart.scss";

const Cart = () => {
  const { products, total, isOpen, openCart, closeCart } = useCart();

  const handleCheckout = () => {
    if (total.productQuantity) {
      alert(
        `Checkout - Subtotal: ${total.currencyFormat} ${formatPrice(
          total.totalPrice,
          total.currencyId
        )}`
      );
    } else {
      alert("Add some product in the cart!");
    }
  };

  const handleToggleCart = (isOpen: boolean) => () =>
    isOpen ? closeCart() : openCart();

  return (
    <div className="cartContainer" style={{ right: isOpen ? "0 " : "-550px" }}>
      <button
        className="cartButton"
        onClick={handleToggleCart(isOpen)}
        style={{
          left: isOpen ? "0 " : "-50px",
          backgroundColor: isOpen ? "#f2ab06ef" : "#002c41",
          transition: "all 0.3s",
        }}
      >
        {isOpen ? (
          <span>X</span>
        ) : (
          <div className="cartIcon">
            <div className="cartQuantity">{total.productQuantity}</div>
          </div>
        )}
      </button>

      {isOpen && (
        <div className="cartContent">
          <div className="cartContentHeader">
            <div className="cartIcon">
              <div className="cartQuantity">{total.productQuantity} </div>
            </div>
            <span className="headerTitle">in Cart</span>
          </div>

          <CartProducts products={products} />

          <div className="cartFooter">
            <p className="sub">SUBTOTAL</p>
            <div className="subPrice">
              <p className="subPriceValue">{`${
                total.currencyFormat
              } ${formatPrice(total.totalPrice, total.currencyId)}`}</p>
              <p className="subPriceInstallment">
                {total.installments ? (
                  <span>
                    {`OR UP TO ${total.installments} x ${
                      total.currencyFormat
                    } ${formatPrice(
                      total.totalPrice / total.installments,
                      total.currencyId
                    )}`}
                  </span>
                ) : null}
              </p>
            </div>
            <button
              className="checkoutButton"
              onClick={handleCheckout}
              autoFocus
            >
              Checkout
            </button>
          </div>
        </div>
      )}
    </div>
  );
};

export default Cart;
