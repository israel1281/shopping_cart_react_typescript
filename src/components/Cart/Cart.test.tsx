import { CartProvider } from "../../contexts/cartContext";

import Cart from ".";

describe("[CartProduct Component", () => {
  const setup = () => {
    return (
      <CartProvider>
        <Cart />
      </CartProvider>
    );
  };

  test("should render correctly", () => {
    const view = setup();
    expect(view).toMatchSnapshot();
  });
});
