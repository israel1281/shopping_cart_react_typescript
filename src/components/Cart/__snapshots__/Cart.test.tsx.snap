// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[CartProduct Component should render correctly 1`] = `
<CartProvider>
  <Cart />
</CartProvider>
`;
